src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -1170,13 +1210,55 @@ int POSIXObject::open(const DoutPrefixProvider* dpp)
   return 0;
 }
 
+int POSIXObject::write_temp_file(const DoutPrefixProvider *dpp)
+{
+  if (obj_fd < 0) {
+    return 0;
+  }
+
+  char temp_file_path[PATH_MAX];
+  // Only works on Linux - Non-portable
+  snprintf(temp_file_path, PATH_MAX,  "/proc/self/fd/%d", obj_fd);
+
+  POSIXBucket *b = dynamic_cast<POSIXBucket*>(get_bucket());
+
+  if (!b) {
+      ldpp_dout(dpp, 0) << "ERROR: could not get bucket for " << get_name() << dendl;
+      return -EINVAL;
+  }
+
+  int ret = linkat(AT_FDCWD, temp_file_path, b->get_dir_fd(dpp), get_temp_fname().c_str(), AT_SYMLINK_FOLLOW);
```

Comment:
You should be able to use linkat(obj_fd, "", b->get_dir_fd(dpp), get_temp_fname().c_str(), AT_EMPTY_PATH); rather than trying to get the path to the tmpfile.  That should link the open FD to the new name.

Reply:
According to the man page for `open` in the section around `O_TMPFILE` it says `AT_EMPTY_PATH` needs the caller to have the CAP_DAC_READ_SEARCH capability to work with linkat(2). 

My understanding is that cap would need to be set on the `radosgw` executable before it's run and that doing so would give the executable wide ranging permissions it didn't have before. Specifically `* Bypass file read permission checks and directory read and execute permission checks;` (https://man7.org/linux/man-pages/man7/capabilities.7.html). Do we set other caps on executables in the Ceph packages like we would have to for this one?

Reply:
I am pretty sure we don't add caps now, but that shouldn't be an obstacle to doing it